course05/homework02/freekpaans/linked_list.h

```diff
@@ -0,0 +1,78 @@
+#include <cstddef>
+#include <string>
+#include <memory>
+#include <iterator>
+
+class LinkedList{
+public:
+	using size_type = size_t;
+
+	struct Link
+	{
+		std::string mValue;
+		std::unique_ptr<Link> mNext;
+	};
+
+	using LinkPtr = std::unique_ptr<Link>;
+
+	LinkedList() =default;
+	LinkedList(const LinkedList& from);
+
+	LinkedList& operator=(const LinkedList& from);
+
+	bool empty() const
+	{
+		return !mFront;
+	}
+
+	size_type size() const;
+
+	void push_front(const std::string& str);
+	void pop_front();
+	void insert_at(size_type index, const std::string& str);
+
+	const std::string& front() const;
+	const std::string& at(size_type index) const;
+
+	struct iterator : public std::iterator<std::input_iterator_tag, std::string, size_type>
+	{
+		const Link* mLink;
```

Comment:
It should be declared at the bottom, but more importantly, be private

Reply:
Why is it important to have it at the bottom?

In this case it was a conscious choice from my side to keep this one as well as `LinkedList::Link` public, because I don't think it hurts much here and would add some code to hide it all.